Union Chat v2

cyyynthia opened the pull request:
It's time to update Union backend owo

This PR is a major rewrite of the API (v2) and a v2 of the websocket! OwO
APIv2 implements more consistent endpoints with cleaner path, a brunch of new features and is covered by unit tests
WebSocket is also in a new version with fancy new features!

**WARNING**: Neither the old API nor the old websocket are supported by this version. Support of the old API is
not complicated, but it's not the case for the current websocket. Partial support would be just a waste of time imo

# Features

 - Clean code (as always)
 - Unit tests `yarn run test` (*note*: you can skip long test with the env variable `SKIP_LONG_TESTS=yes`)
 - Stable and consistent
 - Ez installation system & migrator `yarn run setup`
 - reCAPTCHA support (*should* work, didn't tested and not covered by unit tests, because I don't know how to test it properly)
 - User & server settings (*very* basic for now)
 - Docs! :D

# Ez installer

The installer will prepare the app to start working quickly and easily

 - Build the beautiful es6 code in ugly but (we all hope) working code
 - Migrates the database (no database required, everything will be created)
 - Download & build union-react <!-- react > all --> (union-website can be used but some adjustments will be probably required)

# Endpoints

POST /api/v2/users - Create an user

GET /api/v2/users/self - Get current user
PUT, PATCH /api/v2/users/self - Updates the current user
DELETE /api/v2/users/self - Delete the current user

POST /api/v2/servers - Create a server
PUT, PATCH /api/v2/servers/:serverId - Delete a server
DELETE /api/v2/servers/:serverId/leave - Leave a server
DELETE /api/v2/servers/:serverId - Delete a server

POST /api/v2/servers/:serverId/invites - Create an invite to a server
POST /api/v2/invites/:invite - Accept an invite

POST /api/v2/servers/:serverId/messages - Post a message
PUT/PATCH /api/v2/servers/:serverId/messages/:messageId - Update a message
DELETE /api/v2/servers/:serverId/messages/:messageId - Delete a message

## Not implemented endpoints

GET /api/v2/themes - Get a list of all themes, ordered by votes (with pagination)
POST /api/v2/themes/:themeId/vote - Upvote a theme
DELETE /api/v2/themes/:themeId/vote - Remove the upvote for a theme

POST /api/v2/themes - Create a theme
PUT, PATCH /api/v2/themes/:themeId - Update a theme
DELETE /api/v2/themes/:themeId - Delete a theme

GET /api/v2/developers/applications - Get applications
POST /api/v2/developers/applications - Create an application
PUT, PATCH /api/v2/developers/applications/:applicationId - Update an application
DELETE /api/v2/developers/applications/:applicationId - Delete an application

GET /api/v2/oauth2/authorize - Oauth2 classic routes
POST /api/v2/oauth2/token - Oauth2 classic routes
POST /api/v2/oauth2/revoke - Oauth2 classic routes

# New WebSocket server

Don't be a Discord. Union websocket is Subscription-based, for the happiness of your network and CPU
See the docs for more details

# Some other ideas & stuff to do

 - Snipe but directly in the API (with maybe more advanced behaviour)
 - Channels & DMs
 - Improve scalability of Union

cyyynthia committed:
Better project structure

- Controllers and middlewares separated from route declaration
- es6 imports & Babel to transpile everything
- eslint to have a beautiful code
- ez installer

cyyynthia committed:
nearly perfect

cyyynthia committed:
Setup unit testing

cyyynthia committed:
Database migrations

cyyynthia committed:
Don't spam console

cyyynthia committed:
More tests, beginning of API v2

cyyynthia committed:
User create endpoint

cyyynthia committed:
Reflect sources on the main repo

Will resolve conflicts later, too lazy rn

cyyynthia committed:
Port users and servers to APIv2

cyyynthia committed:
Implement invites

cyyynthia committed:
Finish APIv1 port to v2

cyyynthia committed:
Merge branch 'pr/1'

cyyynthia committed:
Add reCAPTCHA support (closes #8)

cyyynthia committed:
Add developer routes (Not implemented yet)

cyyynthia committed:
Implement user settings (hi #9)

cyyynthia committed:
Implement server settings (closes #9)

cyyynthia committed:
Finally remove support for API v1

I can't find a clean way to support two versions of the websocket
server. Partial support is just a waste of time, so let's do a breaking
update

cyyynthia committed:
Move websocket in a subfolder

cyyynthia committed:
Start some docs (hi #12)

cyyynthia committed:
I guess this is how to use Jekyll

cyyynthia committed:
Socket docs

cyyynthia committed:
Fix package.json

cyyynthia committed:
Start API docs

cyyynthia committed:
Prepare the theme feature

cyyynthia committed:
Finish docs (closes #12) and fix some stuff

cyyynthia committed:
New socket tests (not written yet)

cyyynthia committed:
Socket core

cyyynthia committed:
Remove not core socket tests

These tests are a bit useless. Important tests are in s_core

cyyynthia committed:
Renaming memes

cyyynthia committed:
Fully implement new socket

cyyynthia committed:
Message edit & deletion

cyyynthia committed:
Update Modesta, typo and add auth requirements

cyyynthia committed:
set NODE_ENV in setup, valid JSON in docs

cyyynthia committed:
Use the latest Modesta

cyyynthia committed:
target='_blank'

cyyynthia committed:
Fix installer

cyyynthia committed:
console.log memes

cyyynthia committed:
Fix SERVER_CREATE event

devoxin commented on Configuration.json.example:
```diff
@@ -13,6 +13,12 @@
+        "maxServersPerUser": 5,
+        "maxAppsPerUser": 5
+    },
+    "recaptcha (Remove this entirely if you don't want to setup it)": {
```
Perhaps the note could be moved to a property within this object to save having to edit this line if the user wants recaptcha (mostly a nitpick here)

devoxin commented on docs/_pages/api.md:
```diff
@@ -0,0 +1,262 @@
+|---------------|-----------|---------------------------------------|
+| id            | snowflake | User internal ID                      |
+| username      | string    | Username                              |
+| discriminator | int       | Discriminator fo the user             |
```
```suggestion
| discriminator | int       | Discriminator for the user             |
```

devoxin commented on docs/_pages/api.md:
```diff
@@ -0,0 +1,262 @@
+| discriminator | int       | Discriminator fo the user             |
+| avatarUrl     | string?   | Url of the user's avatar              |
+| servers       | int[]     | List of the server IDs the user is in |
+| online        | boolean   | Whenever or not the user is online    |
```
```suggestion
| online        | boolean   | Whether or not the user is online    |
```

devoxin commented on docs/_pages/api.md:
```diff
@@ -0,0 +1,262 @@
+<span class='api-endpoint'><span class='sun-flower-text'>POST</span> /servers</span>
+<span class='api-auth'>Requires authentication</span>
+
+**Note:** You can't create more server than specified in `maxServersPerUser`
```
```suggestion
**Note:** You can't exceed the `maxServersPerUser` limit
```

devoxin commented on docs/_pages/socket.md:
```diff
@@ -0,0 +1,130 @@
+| Field         | Type         | Description                                                        |
+|---------------|--------------|--------------------------------------------------------------------|
+| id            | snowflake    | User ID                                                            |
+| username      | string       |                                                                    |
```
Missing desc?

Devoxin commented:
this is so damn big.


On another note, please lint with the old settings (semicolons, yoda, etc...)

cyyynthia committed:
eslint changes

cyyynthia committed:
Grammar fixes

Bowser65 commented:
Done :D

cyyynthia committed:
Remove unused eslint dependencies

Bowser65 commented:
oof also just notices some stuff:
 - I completely forgot to impement the rate limit documented in api.md
 - `s_core.js` test is inconsistent and behave weirdly, fails randomly and keeps the process alive, I'll probably remove it too

I'll fix that a bit later and add a comment when it's fixed

cyyynthia committed:
Removed s_core and implemented rate limiting

s_core was inconsistent and didn't behave like a test should do (random
failures and keeps the process alive)
Rate limiting is here to prevent API abuse and is Cloudflare-proof (IP
detection)

Bowser65 commented:
Fixed and changes pushed. Note that the rate limiter should use a Redis server in the future, to improve reliability and scalability

Bowser65 commented:
And on another note, i'm just wondering, why you prefer yoda syntax?

cyyynthia committed:
Let's be nice and unban people
